### ISSUE TITLE
我有几个疑问

### ISSUE THREAD

**Issue by shanniruo** [Tianxiaomo/pytorch-YOLOv4#72]
@Tianxiaomo 1.https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/master/train.py#L171 在这里计算obj_mask时为什么需要用到pred与gt的iou呢？这样岂不是每个iteration obj_mask都会改变？我先前看过别的yolo3代码，obj_mask直接就是这个https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/master/train.py#L177
2.你在计算损失函数时只有obj_mask，计算置信度损失时似乎没有考虑负样本
3.https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/master/train.py#L220 在这里为何使用bceloss呢？论文中说的好像是使用超过1的因子乘以sigmoid，或者是iou损失。我不能理解这并不是一个分类任务使用交叉熵损失有什么优势吗？
以上可能我的理解有误，大佬有空的时候，还请指教一下

**Comment by oreo-lp** [comment]
你好，我也有这方面的疑问。我觉得是不需要pred与gt进行iou来得到obj_mask，它应该由ground truth的来决定。如果这样的话，https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/32abc9266b28031676245450c64a1d140bb3fb9e/train.py#L166从这一行开始到171是不是都不需要了？

**Comment by oreo-lp** [comment]
你好，我想谈一下自己的看法。obj_mask的作用是用来为predictions选择合适的anchor box。v3中每个grid都有三种不同的anchor box，对于包含物体的的grid，我们只能选择三个中最合适的anchor box。衡量合不合适的标准就是predictions与ground truth的IOU，IOU越大的anchor box就是最合适的anchor box。因此，我恰好认为https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/master/train.py#L177 这段代码恰好是用来保护ground truth的anchor box不被消除。因为在https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/32abc9266b28031676245450c64a1d140bb3fb9e/train.py#L216 中，将gt和obj_mask进行相乘了。

但是，我有一点不是很理解的是，在https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/32abc9266b28031676245450c64a1d140bb3fb9e/train.py#L171 中，为什么要进行取反？这里不是很理解。

**Comment by oreo-lp** [comment]
我还有以下几个问题，不知道作者能不能解答，或者哪位同行能够不吝赐教
（1）https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/32abc9266b28031676245450c64a1d140bb3fb9e/train.py#L153 bboxes_iou 我认为函数的一个参数xyxy=Flase。原因如下，这一行代码的作用是用来计算两个bboxes的IOU的，其中有个参数xyxy默认为True，说明这两个bboxes都是以(x1,y1,x2,y2)的进行计算的。但是，在这之前，truth_box和ref_anchors不都是转成(x,yw,h)的形式了嘛，那么xyxy不就应该是False了嘛？

（2）https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/32abc9266b28031676245450c64a1d140bb3fb9e/train.py#L187 我不太理解tgt_scale这个张量的作用是什么？

（3）https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/32abc9266b28031676245450c64a1d140bb3fb9e/train.py#L216 不太理解为什么要对target进行obje_mask, tgt_mask, tgt_scale乘法运算。不是只需要对output进行mask就行了嘛？

**Comment by oreo-lp** [comment]
我好像明白了这一行代码为什么要取反了：
https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/32abc9266b28031676245450c64a1d140bb3fb9e/train.py#L171 
这一行代码要配合下面这一行代码一起看：
https://github.com/Tianxiaomo/pytorch-YOLOv4/blob/32abc9266b28031676245450c64a1d140bb3fb9e/train.py#L177

这两行代码的意思就是说，在训练的过程中，我们只训练那些学习效果不好的bounding box（pred和gt的IOU小于一定阈值的bbox）以及学习那些应该用于预测ground truth的bbox

**Comment by AIMaex** [comment]
Hi,
the reason is written in the YOLOv3 paper (https://arxiv.org/abs/1804.02767) in section 2.1. The bounding boxes which have an IoU with the ground truth > threshold, but are not the best overlapping ones are ignored. The best overlapping ones and the non-overlapping ones are considered for the objectiveness score though.

**Comment by AIMaex** [comment]
@oreo-lp _bboxes_iou 我认为函数的一个参数xyxy=Flase。原因如下，这一行代码的作用是用来计算两个bboxes的IOU的，其中有个参数xyxy默认为True，说明这两个bboxes都是以(x1,y1,x2,y2)的进行计算的。但是，在这之前，truth_box和ref_anchors不都是转成(x,yw,h)的形式了嘛，那么xyxy不就应该是False了嘛？

This is not necessary as x,y are 0 while the function bboxes_iou gets called, so only the width and height are given resulting in the top left corner to be zero and bottom right corner to be the width and height respectively. This means the IoU is the same as with xyxy coordinates given explicitely.

**Comment by zxj11838** [comment]
@ oreo-lp  

Loss includes confidence loss, bounding box loss, and category loss.

obj_mask:  which is the positive and negative sample mask after screening，used to calculate the confidence loss；

tgt_mask: which is the mask of positive samples, used to calculate the bounding box loss and category loss;

tgt_scale: used to balance the loss of large and small targets. Avoid that the error of the large frame has a greater impact on the loss than the error of the small frame.
